### PR TITLE
Use futures to allow concurrent send and receive

### PR DESCRIPTION
**Pull request by SebbyLaw**
### Description

Rewrite the logic of the client (and server) to allow for concurrent sends and receives.
Resolves the issue found by several library users when calling `Client.request(...)` very quickly:
```
  File "/projectpath/venv/lib/python3.9/site-packages/discord/ext/ipc/client.py", line 104, in request
    recv = await self.websocket.receive()
  File "/projectpath/venv/lib/python3.9/site-packages/aiohttp/client_ws.py", line 216, in receive
    raise RuntimeError("Concurrent call to receive() is not allowed")
RuntimeError: Concurrent call to receive() is not allowed
```

### Checklist

- [ ] This PR has been tested.

### CODE REVIEW

**Comment by SebbyLaw** [comment]
Oops, meant to open as a draft and hit enter

**Comment by SebbyLaw** [comment]
Ready for review I guess. This still hasn't been tested.

File: discord/ext/ipc/client.py [hunk 1]
```diff
@@ -103,57 +110,98 @@ async def init_sock(self):
...
+        asyncio.TimeoutError
+            If a timeout is provided and it was reached.
         """
         log.info("Requesting IPC Server for %r with %r", endpoint, kwargs)
         if not self.session:
             await self.init_sock()
 
+        # generate a unique identifier for this request
+        # there is probably a better way to do this
+        nonce = uuid.uuid1()
```
Comment (bijij) [inline discord/ext/ipc/client.py:145]:
```suggestion
        nonce = os.urandom(16).hex()
```

Personally i'd use something like this.

```py
In [1]: %timeit uuid.uuid1()
1.98 µs ± 9.62 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [2]: %timeit os.urandom(16).hex()
130 ns ± 0.409 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

You'll need to change the imports though.

File: discord/ext/ipc/client.py [hunk 2]
```diff
@@ -82,57 +89,98 @@ async def init_sock(self):
...
         """Make a request to the IPC server process.
 
         Parameters
         ----------
         endpoint: str
             The endpoint to request on the server
+        timeout: float
+            The number of seconds to wait for a response from the server
+            before timing out and raising :exc:`asyncio.TimeoutError`.
+            By default this is ``None`` and waits forever.
```
Comment (ShineyDev) [inline discord/ext/ipc/client.py:111]: I'm not sure defaulting to waiting indefinitely is a great idea.

**Review by ShineyDev — COMMENTED** [review]
I'll let @lgaan review the content of this PR.

**Pull request closed without merging**